elm/benchmark_diff.py

```diff
@@ -0,0 +1,161 @@
+"""
+This is the benchmark test of diff models.
+"""
+import functools
+import itertools
+import re
+import hydra
+import torch
+import multiprocessing as mp
+from tqdm import tqdm
+from typing import Union
+from omegaconf import OmegaConf
+
+from elm.codegen.codegen_utilities import set_seed
+from elm.constants import SRC_PATH
+from transformers import AutoTokenizer, AutoModelForCausalLM, AutoConfig
+from benchmarks import parity_reference, eval_code_string
+from elm.util.diff_eval import split_diff, apply_diff
+
+
+parity_test_data = {
+            i: parity_reference(*i) for i in itertools.product(range(2), repeat=4)
+        }
+
+
+def eval_completions(
+    codes: Union[str, list[str]], task: str = "parity", timeout: int = 5
+) -> Union[int, list[int]]:
+    """
+    Evaluate (a batch of) the modified codes on a task.
+    Args:
+        codes: either a string or a batch of strings. The code(s) to be evaluated.
+        task: (Optional) the task to be performed.
+        timeout: (Optional) the timeout (in seconds).
+    Returns:
+        either the status code of the string or a list of status codes of the batch
+        of strings (depending on whether `codes` is batched).
+    """
+    if task == "parity":
+        _codes = codes if isinstance(codes, list) else [codes]
+        results = []
+        for code in _codes:
+            results.append(eval_code_string(code, parity_test_data, timeout))
+        if isinstance(codes, list):
+            # Batch evaluation returns the batch
+            return results
+        else:
+            # Single evaluation returns a single result
+            return results[0]
+    else:
+        raise ValueError(f"Unknown task: {task}")
+
+
+def mutate_code(n_bugs: int = 5, task: str = "parity") -> tuple:
+    """
+    Mutate code to create n bugs. Output the prompt in diff format.
+    Args:
+        n_bugs: number of bugs to introduce (from 1 to 5).
+        task: (Optional) the task to be performed.
+    Returns:
+        mutated_code, function_string
+    """
+    mutation_template = [
+        f"<NME> {task}.py\n<BEF> ",
+        "\n<MSG> Fixed bugs",
```

Comment:
Are we not actually appending <DFF> to the prompt here?

Reply:
This was modified based on `benchmark.py` and I got confused about this line too (where was the code??) :joy:. Is this what you are asking? Turns out there is a `mutation_template.insert(1, func_str)` in line 76 that inserts the code to the middle. But maybe we can make it less confusing.

Reply:
Yeah I think I wrote the insert and now I've been confused by my own inherited code 😆